### ISSUE TITLE
performance drop from v1.1.8 to >= v1.2.0 when using metrics

### ISSUE THREAD

**Issue by egonuel** [PyTorchLightning/pytorch-lightning#6605]
## 🐛 Bug

Hi, I'm observing longer epoch times for pytorch lightning versions >= 1.2.0 when using metrics in the training or validation steps. Depending on the model, dataset etc I observed an increase of 10 seconds per epoch.
I tested it using Accuracy() and ConfusionMatrix() and the impact is larger for the former.
Removing the metrics results in the same epoch times regardless of the lightning version.
My colleagues and I verified this on different machines using a simple mnist model.
This behavior also occurs with the soon-to-be deprecated internal metrics and the torchmetrics variant of Accuracy()
Currently, the mnist host website is down, so I replicated this using cifar10 in the boring model notebook, to which a link can be found below.
The impact in this example is not as severe as observed with our mnist example, but still in the range of 3-5 seconds per epoch on my machine, which some might find neglectable, but this adds up.
The problem using colab is, that I was not able to install pytorch lightning 1.1.8, since it is incompatible with the current torchtext library, which is imported by lightning if available, and is installed in the colab environment by default. Uninstalling it using pip did not work. To verify this one unfortunately has to run this somewhere else
The printed epoch/fit times in the notebook where only measured using 1.2.4.
The following times for 3 epochs and the complete fit were measured on my machine using a V100 for 1.1.8 and 1.2.4:

```
pl 1.2.4:
epoch: 64 s
epoch: 64 s
epoch: 64 s
fit: 198 s
```

```
pl 1.1.8
epoch: 59 s
epoch: 58 s
epoch: 59 s
fit: 182 s
```

I verified this behavior for all versions since 1.2.0 including the 1.2.0rcs.

## Please reproduce using the BoringModel

[https://colab.research.google.com/drive/1ILDOsZqlOKplrdKBkuGUiNiWnJwYYYoW?usp=sharing](https://colab.research.google.com/drive/1ILDOsZqlOKplrdKBkuGUiNiWnJwYYYoW?usp=sharing)

### To Reproduce

Use following [https://colab.research.google.com/drive/1ILDOsZqlOKplrdKBkuGUiNiWnJwYYYoW?usp=sharing](https://colab.research.google.com/drive/1ILDOsZqlOKplrdKBkuGUiNiWnJwYYYoW?usp=sharing)

### Expected behavior

Same epoch times regardless of lightning version

### Environment

this is the machine used for the times above

* CUDA:
	- GPU:
		- Tesla V100-SXM2-32GB
		- Tesla V100-SXM2-32GB
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.20.1
	- pyTorch_debug:     False
	- pyTorch_version:   1.8.0
	- pytorch-lightning: 1.2.4
	- tqdm:              4.59.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.8.8
	- version:           #141-Ubuntu SMP Fri Feb 19 13:46:27 UTC 2021

**Comment by tchaton** [comment]
@SkafteNicki Any ideas ?

**Comment by SkafteNicki** [comment]
I can explain this. From v1.2 and accuracy metric gained extra features such as support for `top_k` calculations, multi-dim tensors and more. Because of all this a lot more input checking is going on behind the scene. Basically we moved from using this simple function:
https://github.com/PyTorchLightning/pytorch-lightning/blob/e429f97b670846bf0e0595074c4d3c4f152dcf27/pytorch_lightning/metrics/utils.py#L46
to this much more advanced version:
https://github.com/PyTorchLightning/pytorch-lightning/blob/e429f97b670846bf0e0595074c4d3c4f152dcf27/pytorch_lightning/metrics/classification/helpers.py#L302
so it is not unthinkable that over a hole epoch the time consumption goes up because we are actually going more in the background.
I am going to mark this as "working as intended", but if there is a general consensus that this is unwanted, I can try to cut the input checking down.

**Comment by carmocca** [comment]
It'd be nice if you could check your hypothesis by benchmarking the `_input_format_classification` function in both versions

**Comment by SkafteNicki** [comment]
Benchmark (time of 1000 evaluations):
New input checker: 0.9793810844421387
Old input checker: 0.1436159610748291
Approx 0.97 / 0.14 = 6.92 times slower.

**Comment by egonuel** [comment]
I also played with the two versions of `_input_format_classification` and this is indeed the cause for the slow-down, the `Accuracy.update()` becomes 10 times slower on my machine and example.
I'm wondering if maybe these checks, and some of them are expensive, can be avoided in the hot batch loop on every input. 
Not that I can put any work in this, but maybe a strict mode to turn the checking on and check it like with `fast_dev_run `in the beginning could be an option.

**Comment by maximsch2** [comment]
Most metrics auto-magically figure out if the thing is binary, multiclass, multi-dim, etc and this of course has a cost (both in compute and in code complexity). I wonder if we should instead take an extra parameter during `__init__` to tell us what we expect to have and then just roll with it? We can additionally check those in `fast_dev_run` as @egonuel is suggesting (and maybe additionally check on first invocation).

**Comment by SkafteNicki** [comment]
Okay, just a bit follow up. When evaluating the the new `_input_format_classification` function the time is approximately spend on:
* 44% on the `select_topk` where nearly all that time goes to calling `torch.topk`
* 39% on the `_check_classification_inputs` function. Here we can probably do a bit to cut the time, but not much
* remaining 17% is spend on various stuff 
So much of the time has to do with the top k calculation. We could differentiate between using `argmax` when `k=1` and `topk` when `k>1` which probably will give performance.
